runtime/vm/LayoutFFITypeTable.cpp

```diff
@@ -0,0 +1,136 @@
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] https://openjdk.org/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
+ *******************************************************************************/
+
+#include "j9.h"
+#include "j9protos.h"
+
+extern "C" {
+
+#if JAVA_SPEC_VERSION >= 16
+
+/**
+ * @brief Return the hash value for an entry in the layout string table.
+ *
+ * @param entry the layout string entry
+ * @param userData the data specified as userData when the hashtable was created,
+ *                 in our case pointer to the port library
+ * @returns the hash value
+ */
+static UDATA
+layoutStrFFITypeHashFn(void *entry, void *userData)
+{
+	J9LayoutStrFFITypeEntry *layoutStrEntry = (J9LayoutStrFFITypeEntry *)entry;
+
+	return computeHashForUTF8(layoutStrEntry->layoutStr, layoutStrEntry->layoutStrLength);
+}
+
+/**
+ * @brief Compare two layout string entries for equality.
+ *
+ * @param leftEntry the first entry to compare
+ * @param rightEntry the second entry to compare
+ * @param userData the data specified as userData when the hashtable was created,
+ *                 in our case pointer to the port library
+ * @returns 1 if the entries match, otherwise return 0
+ */
+static UDATA
+layoutStrFFITypeHashEqualFn(void *leftEntry, void *rightEntry, void *userData)
+{
+	J9LayoutStrFFITypeEntry *leftLayoutStrEntry = (J9LayoutStrFFITypeEntry *)leftEntry;
+	J9LayoutStrFFITypeEntry *rightLayoutStrEntry = (J9LayoutStrFFITypeEntry *)rightEntry;
+
+	return J9UTF8_DATA_EQUALS(leftLayoutStrEntry->layoutStr, leftLayoutStrEntry->layoutStrLength,
+			rightLayoutStrEntry->layoutStr, rightLayoutStrEntry->layoutStrLength);
+}
+
+/**
+ * @brief Delete an entry from the layout string hashtable.
+ *
+ * @param entry the entry to be deleted
+ * @param userData the data specified as userData when the hashtable was created,
+ *                 in our case pointer to the port library
+ * @return TRUE after the entry is deallocated
+ */
+static UDATA
+freeLayoutStrFFIDoFn(void *entry, void *userData)
+{
+	PORT_ACCESS_FROM_PORT((J9PortLibrary *)userData);
+	J9LayoutStrFFITypeEntry *layoutStrEntry = (J9LayoutStrFFITypeEntry *)entry;
+	j9mem_free_memory(layoutStrEntry->layoutStr);
+	j9mem_free_memory(layoutStrEntry);
+	return TRUE;
+}
+
+/**
+ * @brief Create the layout string hashtable.
+ *
+ * @param vm a pointer to J9JavaVM
+ * @return the hasthable
+ */
+J9HashTable *
+createLayoutStrFFITypeTable(J9JavaVM *vm)
+{
+	return hashTableNew(OMRPORT_FROM_J9PORT(vm->portLibrary), "Layout string table", 0,
+			sizeof(J9LayoutStrFFITypeEntry), 0, 0, J9MEM_CATEGORY_VM_FFI, layoutStrFFITypeHashFn, layoutStrFFITypeHashEqualFn, NULL, NULL);
```

Comment:
This should be split further - perhaps even one argument per line.

Reply:
Fixed.